### PR TITLE
Use contexts to redefine units

### LINKED ISSUE #853: Context.define for currency conversions
I'm trying to use pint to perform currency conversions.
Currency FX rates change by the minute and may be different in different markets/situations - which would be modelled perfectly by a pint.Context.

However, while Context allows defining custom conversion functions between different dimensions, it doesn't seem to allow customizing conversions within the same dimension.

I'm currently using throwaway UnitRegistry objects (very expensive to create):
```python
import pint

ureg_day1 = pint.UnitRegistry()
ureg_day1.define("USD = [currency]")
ureg_day1.define("EUR = 1.11 USD")
ureg_day1.define("GBP = 1.22 USD")

ureg_day2 = pint.UnitRegistry()
ureg_day2.define("USD = [currency]")
ureg_day2.define("EUR = 1.13 USD")
ureg_day2.define("GBP = 1.28 USD")

print((100 * ureg_day1("EUR")).to(ureg_day1("GBP")))
print((100 * ureg_day2("EUR")).to(ureg_day2("GBP")))
```
output:
```
90.98360655737706 GBP
88.28124999999999 GBP
```

I could not find a way to convert 100 EUR to USD on day1 and then convert them back to EUR on day2. This doesn't work:
```python
x = 100 * ureg_day1.EUR
x = x.to(ureg_day1.USD)
x = x.to(ureg_day2.EUR)
x
```
Output:
```
100.0 EUR  # Wrong - it should be 98.23 EUR!
```
Workaround:
```python
x = 100 * ureg_day1.EUR
x = x.to(ureg_day1.USD)
x = x.magnitude * ureg_day2.USD
x = x.to(ureg_day2.EUR)
x
```
Output:
```
98.23008849557525 EUR
```

It would make sense to have instead something like:
```python
import pint

ureg = pint.UnitRegistry()
ureg.define("USD = [currency]")
ureg.define("EUR = <TODO: valid currency, but crash if you attempt conversion>")
ureg.define("GBP = <TODO: valid currency, but crash if you attempt conversion>")

ctx_day1 = pint.Context("")
ctx_day1.define("EUR = 1.11 USD")
ctx_day1.define("GBP = 1.22 USD")

ctx_day2 = pint.Context("")
ctx_day2.define("EUR = 1.13 USD")
ctx_day2.define("GBP = 1.28 USD")

x = 100 * ureg.EUR
print(x.to(ureg.GBP, ctx_day1))
print(x.to(ureg.GBP, ctx_day2))

# x = x.to(ureg.USD)  # crash
x = x.to(ureg.USD, ctx_day1)
x = x.to(ureg.EUR, ctx_day2)
print(x)
```
Output:
```
90.98360655737706 GBP
88.28124999999999 GBP
98.23008849557525 EUR
```

### LINKED ISSUE #932: Disable all throwaway contexts
Anonymous contexts (#870) are enabled on the fly with ``ureg.enable_contexts()``, but are not previously registered with ``ureg.add_context()``. Regardless of having a name or not, it is convenient in some cases to use throwaway contexts.

When there are more throwaway contexts enabled than the number of contexts permanently registered with ``ureg.add_context()``,  then ``ureg.disable_contexts()``  invoked without parameter fails to disable them all.

```python
>>> import pint
>>> ureg = pint.UnitRegistry()
>>> len(ureg._active_ctx.maps)
1  # Empty context - this is a separate bug
>>> ureg.enable_contexts(*(pint.Context() for _ in range(len(ureg._contexts) + 3)))
>>> len(ureg._active_ctx.maps)
15
>>> ureg.disable_contexts()
>>> len(ureg._active_ctx.maps)
4
>>> ureg.disable_contexts()
>>> len(ureg._active_ctx.maps)
0
```
The context manager syntax is unaffected:
```python
>>> with ureg.context(*(pint.Context() for _ in range(len(ureg._contexts) + 3))):
...     pass
>>> len(ureg._active_ctx.maps)
0
```

### PR DESCRIPTION
**Pull request by crusaderky**
- [x] Closes #853 - Context.define for currency conversions
- [x] Closes #932 - Disable all throwaway contexts
- [x] Contexts that only implement transforms between dimensionalities (that is, all contexts until now) no longer require their own cache and therefore are enabled in microseconds even on the first time; contexts that implement redefinitions require a *much* smaller cache than before which is very quickly built on demand.
- [x] NaN is now parsed as a float

- [x] Executed ``black -t py36 . && isort -rc . && flake8`` with no errors
- [x] The change is fully covered by automated unit tests
- [x] Documented in docs/ as appropriate
- [x] Added an entry to the CHANGES file

# Demo
```python
import pint

ureg = pint.UnitRegistry()
ureg.load_definitions("""
USD = [currency]
EUR = NaN USD
GBP = NaN USD

@context currencies_20191218
    EUR = 1.11 USD
    GBP = 1.18 EUR
@end

@context dungeons_and_dragons = dnd
    # Simplified conversion rates for the
    # non-English versions of the D&D manuals
    pound = 0.5 kg
    gallon = 4 l
    foot = 1.5/5 m
@end
""".splitlines())

>>> ureg.GBP.dimensionality
<UnitsContainer({'[currency]': 1.0})>
>>> ureg.Quantity("10 GBP")
10 GBP
>>> ureg.Quantity("10 GBP").to("USD")
nan USD
>>> ureg.Quantity("10 GBP").to("USD", "currencies_20191218")
13.098 USD
>>> ureg.Quantity("10 ft").to("m")
3.0479999999999996 meter
>>> ureg.Quantity("10 ft").to("m", "dnd")
3.0 meter
```

### CODE REVIEW

File: pint/registry.py [hunk 1]
```diff
@@ -768,7 +752,7 @@ def get_base_units(self, input_units, check_nonmult=True, system=None):
         return self.get_root_units(input_units, check_nonmult)
 
     def _get_root_units_recurse(self, ref, exp, accumulators):
-        for key in sorted(ref):
+        for key in ref:
```
Comment (crusaderky) [inline pint/registry.py]: Couldn't find any reason for sorting?

File: pint/registry.py [hunk 2]
```diff
@@ -887,6 +883,25 @@ def parse_unit_name(self, unit_name, case_sensitive=True):
...
+        candidates = dict.fromkeys(candidates)  # ordered set
+        for cp, cu, cs in list(candidates):
+            assert isinstance(cp, str)
+            assert isinstance(cu, str)
+            if cs != "":
+                raise NotImplementedError("non-empty suffix")
+            if cp:
+                candidates.pop(("", cp + cu, ""), None)
+        return tuple(candidates)
+
```
Comment (crusaderky) [inline pint/registry.py]: Worth noting that this function was not doing what its docstring said before; it was deduplicating only identical tuples.

File: pint/registry.py [hunk 3]
```diff
@@ -1400,6 +1485,7 @@ def _get_compatible_units(self, input_units, group_or_system):
         ret = super()._get_compatible_units(input_units, group_or_system)
 
         if self._active_ctx:
+            ret = ret.copy()  # Do not alter self._cache
```
Comment (crusaderky) [inline pint/registry.py]: This line missing was the one and only reason why contexts required their own cache before!

**Comment by crusaderky** [comment]
@jetheurer could you play with this and tell me how it goes?

**Comment by crusaderky** [comment]
Changed undef to NaN. The user will have to implement his own ``math.isnan`` tests.
Couldn't find a simple way to tell apart internal transformation performed by pint from the final operations performed by the user.

**Comment by crusaderky** [comment]
@hgrecco ready for review and merge :sunglasses:

**Comment by hgrecco** [comment]
Great PR! It adds functionality and improves the existing one.

Just as a closing note: Now that unit definitions are possible within contexts, using currencies is possible. However, I would like to keep pint within the physical units realm. More involved currency conversion features such as downloading values from the web for certain dates, etc would be welcomed as an external package.

**Comment by crusaderky** [comment]
Downloading currency fx rates from the Internet was never in my plans. Pandas-datareader already does it well enough for any prototype work, and writing a function that generates a pint context out of its dataframes is trivial. For production work, each company will have its own bespoke data source, so any integration has to be bespoke as well.

**Comment by jetheurer** [comment]
I love it!  Thanks @crusaderky

**Comment by numpde** [comment]
This looks quite useful, thank you. I'd like to step it up a notch and have the same currency on different dates (or even timestamps) behave essentially like different units (because of the distinction of 'nominal' and 'real' value). Is there a slicker way than defining a bunch of units like EUR_20200515, EUR_20200516, USD_20200515, ...?

**Comment by crusaderky** [comment]
@numpde no there isn't. May I ask why you want to do that? I'm using pint for a trading platform and I am pretty satisfied with modelling FX with one context per timepoint (day or intraday). And I can use transforms to implement asymmetric curves, FX commissions, etc.

**Comment by numpde** [comment]
@crusaderky In principle, currency denominations across time shouldn't mix for the same reason that different currencies don't: their value is nominal. In the long term this is usually because of inflation. In financial projections, cash flows at different timepoints are discounted to a reference timepoint (e.g. the present) before adding them up to capture the [time value of money](https://www.investopedia.com/terms/t/timevalueofmoney.asp). So, while we're all "[consenting adults](https://pint.readthedocs.io/en/latest/defining.html)", I'm looking for a currency unit framework/library that would encourage this distinction (and would also digest currency conversion at time-variable rates, and interoperate with pandas/numpy).

**Comment by crusaderky** [comment]
@numpde I'm well aware of how discounting and Mark-to-Market pricing works. Still don't get why the current design (one context per timestamp) wouldn't allow you to model it.

It's worth noting that you can pass a Quantity that wraps a numpy vector to a transform; this could allow you to model a time axis and/or deterministic or Monte Carlo shocks on your FX rates. Sadly labelling with xarray is fidgety today - see discussion at https://github.com/hgrecco/pint/issues/1099

**Pull request merged**